### ISSUE TITLE
MAJOR Security Issue - do NOT use this container until this is fixed

### ISSUE THREAD

**Issue by alexandery** [dperson/openvpn-client#189]
This container exposes ALL child containers to the outside world. It is NOT safe to use it. So, if you think -f flag works - you are wrong.

IPTables configuration (firewall settings) is completely wrong:
- allows ALL incoming connections from anywhere into your machine (think - no firewall at all!)
- does NOT implement a kill switch - all of your containers will connect using host network when VPN drops, thus making this setup completely useless

Easy to test:
- Start your setup
- Get your VPN external IP
  - One way:
    - "docker exec -ti vpn_container_name /bin/bash"
    - "curl -s http://ifconfig.me"
  - Another way:
    - "docker logs vpn_container_name" - you will see full connection log, and at the bottom of it there are couple of lines that will have your public IP
- Open a browser and navigate to your external IP and any app port - you'll be presented with its web UI exposed to the whole world.
- Or use "nmap -Pn IP-Address" command from Linux command line to see all open ports. You'll be surprised.

I'm wondering as to why even the most basic IPTables rule to protect you is not there?

Also, firewall should be always engaged. Why is it optional - makes no sense.

**Comment by steamyRayV** [comment]
This was exactly my concern but I was hoping I was misunderstanding a concept. In cases where people are using a shared VPN server, this doesn't expose web services unless they also set the port to forward in the user console of the VPN provider. I still see that as a big issue because it places unneeded trust in the VPN provider's security. In your case it sounds like you have your own VPN server that is forwarding everything? That's a nightmare.

**Comment by steamyRayV** [comment]
What prompted you to test this on your server? Was it my post from the other day? I had been using it for a while without considering this.

**Comment by alexandery** [comment]
Not sure what your vpn provider does, but as long as you've got public IP via your provider - you are exposed. I presume you get public IP.

I've tested it with a vpn provider, not my own vpn server. I first noticed that in my tests my child containers were able to access internet when i forced vpn to fail. I started digging deeper when I realized all child containers are exposed. Then i started looking into what's going on inside the container and found that IPTables configuration is so weird it doesn't make much sense.

What's surprising is that so many people recommend this docker image, yet they either never really test what it does or... don't know..

**Comment by steamyRayV** [comment]
I tested it 2 years ago and the firewall did work back then. It has been a while though.
I think my external IP is shared with other people, so it the VPN server must know who to forward a specific port to. I'd have to forward the port of my web service to expose them to the web. Even so, I agree that there needs to be a firewall. It was many-many versions ago when I tested to see if it was blocking all traffic from going outside the VPN.

**Comment by das-peter** [comment]
Looks like @davebv is working on some iptables adjustments: https://github.com/davebv/openvpn-client/commit/a3b49fd5dafda50a358432122562c8997714363d

Maybe we can get this peer-reviewed and merged if suitable?

**Comment by alexandery** [comment]
> Looks like @davebv is working on some iptables adjustments: [davebv@a3b49fd](https://github.com/davebv/openvpn-client/commit/a3b49fd5dafda50a358432122562c8997714363d)
> 
> Maybe we can get this peer-reviewed and merged if suitable?

I'm not great with iptables (I really suck, tbh). However I long ago adjusted my copy of this image and stopped all leaks. I've also removed bunch of superfluous statements from .sh file and made it work for my provider without making it extensible for any other. So my version may work for some but not for others.

The most important statement was never in the original (never in it's history): iptables -P INPUT DROP , so anyone using this image was *always* exposing their containers to the outside. Also, this image always allowed traffic to go through host network when vpn dropped. Pretty much it was useless. I don't understand how others didn't see that (or nobody ever tested?).

The following I've tested with failing vpn, made sure no ports are exposed and local network able to access the containers.

```
    iptables -P INPUT DROP
    iptables -P OUTPUT DROP
    iptables -P FORWARD DROP

    iptables -A OUTPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT
    iptables -A INPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT

    # Required to establish VPN connection
    iptables -A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT

    # Allow loopback interface to do anything
    iptables -A INPUT -i lo -j ACCEPT
    iptables -A OUTPUT -o lo -j ACCEPT

    # Allow access to DNS
    iptables -A OUTPUT -d #MY_DNS_SRV1 -j ACCEPT
    iptables -A OUTPUT -d #MY_DNS_SRV2 -j ACCEPT

    iptables -A OUTPUT -p tcp -m tcp --dport 53 -j ACCEPT
    iptables -A OUTPUT -p udp -m udp --dport 53 -j ACCEPT
    iptables -A INPUT -p tcp -m tcp --dport 53 -j ACCEPT
    iptables -A INPUT -p udp -m udp --dport 53 -j ACCEPT

    iptables -A OUTPUT -o tun+ -j ACCEPT
    iptables -A OUTPUT -d ${docker_network} -j ACCEPT

    iptables -A OUTPUT -p udp -m udp --sport 1194 -j ACCEPT

    iptables -A OUTPUT -p udp -m udp --dport 1194 -j ACCEPT
    iptables -A OUTPUT -p tcp -m tcp --dport 1194 -j ACCEPT

```

**Comment by das-peter** [comment]
Unfortunately I suck at iptables too :P But maybe we can combine our no-ledge to knowledge...

If I read the changes from @davebv correctly the code should generate iptables like this:
```
  iptables -F
  iptables -X
  # Default filter policy
  iptables -P INPUT DROP
  iptables -P INPUT DROP
  iptables -P FORWARD DROP
  # Accept on localhost
  iptables -A INPUT -i lo -j ACCEPT
  iptables -A OUTPUT -o lo -j ACCEPT
  # Allow established sessions to receive traffic
  iptables -A INPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT

  iptables -A FORWARD -j ACCEPT
  iptables -A OUTPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT
  iptables -A OUTPUT -o tap0 -j ACCEPT
  iptables -A OUTPUT -o tun0 -j ACCEPT
  iptables -A OUTPUT -d ${docker_network} -j ACCEPT

  iptables -A OUTPUT -p udp -m string --hex-string "${rr}" --algo bm -m udp --dport 53 -j ACCEPT

  iptables -A OUTPUT -p tcp -m owner --gid-owner vpn -j ACCEPT 2>/dev/null &&
  iptables -A OUTPUT -p udp -m owner --gid-owner vpn -j ACCEPT || {
    iptables -A OUTPUT -p tcp -m tcp --dport $port -j ACCEPT
    iptables -A OUTPUT -p udp -m udp --dport $port -j ACCEPT; 
  }
```

Looks pretty similar to what @alexandery posted. With more placeholders and some differences:

1.   `iptables -F` & `iptables -X` - if I understand correctly this cleans the slate before beginning. Should be ok.
2. `${IPT} -P INPUT DROP` is duplicated so I guess that should be `${IPT} -P OUTPUT DROP` instead
3. `iptables -A FORWARD -j ACCEPT` Not sure what this is for.
4. Access to DNS is based on dynamic created rule: `  iptables -A OUTPUT -p udp -m string --hex-string "${rr}" --algo bm -m udp --dport 53 -j ACCEPT`

From what I understand we should fix. 2. and probably figure out 2. but other than that the generic approach from @davebv seems to be a match.
Next question would be if we can create a small test-suite to check the behavior.

**Comment by das-peter** [comment]
The created pull request #192 should be considered work in progress.

**Comment by alexandery** [comment]
In my tests anything with ```-m owner --gid-owner``` allows child containers to connect out when vpn is down. I've explicitly tested with and without that. Also, this is a very unusual statement (I haven't run into it anywhere in my long research into iptables) so I did not trust it.

I'm not sure about dynamic DNS. Not sure why, but it didn't work for me correctly (maybe there was dns leak? i don't remember from my tests now back then), but I scraped that and hardcoded known dns servers. You need to be able to resolve vpn server, so this statement will default to use whatever DNS entries were *prior* to establishing vpn connection. And you don't want to use anything but vpn provided dns resolution.

Pretty much, iptables rules should be small and contain only what is required. Anything extra, that doesn't make sense, shouldn't be there. So, things like ```iptables -A FORWARD -j ACCEPT``` or ```iptables -A OUTPUT -o tap0 -j ACCEPT``` shouldn't be there.

To be honest, I have my reservation about this image altogether. I find it *highly* unlikely that someone would simply miss ```iptables INPUT DROP``` - this is the first thing you think of when working with firewalls. That's why I've created my own image I use now and I've looked through the .sh file.

All in all this image has way too many leaks / problems to be used for what it was supposed to be used.

**Comment by davebv** [comment]
Disclaimer, my changes https://github.com/davebv/openvpn-client/commit/a3b49fd5dafda50a358432122562c8997714363d are WIP.

`-A FORWARD ACCEPT` should allow  forwarding the traffic, missing the interface (should it be added?).
These are snippets that could help:
https://www.linode.com/docs/networking/vpn/tunnel-your-internet-traffic-through-an-openvpn-server/
https://gist.github.com/Tristor/ed0f6867d2b0fa4c1f80300af6e0e12e

**Comment by rayban099** [comment]
I'm just throwing my two cents out there.  I came here because I was having the same issues others have identified.  

- The IP address returned by curl was not that of the vpn.  My connection was exposed
- The vpn connection reset every 5 seconds due to no activity.

TL;DR - Compare a working OpenVPN config file with the one generated by the image.  If they are different, the image version can be rectified by updating openvpn.sh within the vpn container.  For more details, read below.

My fix was very simple.  I compared the OpenVPN config file generated by the image with one that worked outside of the container.  A significant amount of configuration was missing (e.g. client certificate and key were missing, flags required by the VPN server were missing or incorrect).  To fix all of this, I modified the openvpn.sh file (/usr/bin/openvpn.sh) within the container.  

The file performs several functions, including pushing out the firewall changes to iptables and generating the OpenVPN config file.  I only modified the portion related to the OpenVPN config file so that the result aligns with a known working config file.  Once this was completed, the image works perfectly.  Containers I configure as dependent on the vpn container no longer expose my data.  I haven't tested whether the linked containers will expose data if the vpn connection is interrupted, yet.  But, otherwise, I'm satisfied.

NOTE: I haven't shared my config files as I lost the original, problematic one.  But, I plan on recreating the issue and sharing the files.

**Comment by alexandery** [comment]
@rayban099 In my case I always used the .conf file and never relied on the generated version.

The major issue is that the firewall configuration is a total mess, that simply doesn't work as it should.

**Comment by dperson** [comment]
1) I don't see the issue that you're reporting, my containers aren't accessible on the IP from my VPN with the open port.
2) I've added a firewall rules to prohibit connections in, except those specifically allowed by `VPNPORT` or `-p`.

**Comment by alexandery** [comment]
You've got NO protection from any incoming connection in your firewall rules! Great example of  "it works for me", hahaha! 

IPTables 101: 
```
iptables -P INPUT DROP
```
which you NEVER had in your rules. Wow.. Just wow..

**Comment by steamyRayV** [comment]
I can confirm that this update did block a port I had been using all along and I had to use the 'VPNPORT' variable to open it back up. Seems like this is corrected!
Thanks for maintaining this container, dperson. Glad you are back!

**Comment by jkaberg** [comment]
@dperson I still see open ports that I havent allowed. With following default web ui port 8112 and torrent port is still open.

```
  pirate_vpn:
    image: dperson/openvpn-client
    restart: unless-stopped
    stdin_open: true
    tty: true
    security_opt:
      - label:disable
    cap_add:
      - net_admin
    read_only: true
    tmpfs:
      - /run
      - /tmp
    volumes:
      - /dev/net:/dev/net:z
      - ${STORAGE_DIR}/vpn:/vpn
    dns:
      - 1.1.1.1
      - 8.8.8.8
    networks:
      - majogw
      - plex

  deluge:
    image: linuxserver/deluge
    restart: unless-stopped
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - ${STORAGE_DIR}/deluge/config:/config
      - /media/storage:/storage
    environment:
      - TZ=Europe/Oslo
      - PUID=911
      - PGID=911
    network_mode: "service:pirate_vpn"
```
